### PR TITLE
fix(ui): fix drag-and-drop file feature for macOS terminal

### LINKED ISSUE #5729: Adding files to CLI within cursor/vscode.
### What happened?

Often I want to quickly add multiple files to CLI context when using it from within cursor. One way to do this would be to drag the file from the left side panel to the terminal where CLI is operating. However, when I drag and drop files, the file paths are often truncated. This does not happen in Claude Code or just terminal. I am attaching an example below. Only part of the file path is added. 

Image

### What did you expect to happen?

I expected the whole file path to be added.

### Client information

│ About Gemini CLI                                            │
│                                                             │
│ CLI Version          0.1.18                                 │
│ Git Commit           cd7e60e0                               │
│ Model                gemini-2.5-pro                         │
│ Sandbox              no sandbox                             │
│ OS                   darwin                                 │
│ Auth Method          OAuth   

### Login information

_No response_

### Anything else we need to know?

_No response_

### LINKED ISSUE #6386: Unable to drag markdown files to CLI within cursor
### What happened?

I can't drag markdown files to CLI.

Image

About Cursor
---
version: 1.4.5
submit: 1.99.3
date: af58d92614edb1f72bdd756615d131bf8dfa5290
Electron: 2025-08-13T02:08:56.371Z
ElectronBuildId: 34.5.8
Chromium: undefined
Node.js: 132.0.6834.210
V8: 20.19.1
OS: 13.2.152.41-electron.0

### What did you expect to happen?

I hope markdown file can also be dragged to CLI.

### Client information

* **CLI Version:** 0.1.21
* **Git Commit:** 055fe46d
* **Operating System:** darwin v22.15.0
* **Sandbox Environment:** no sandbox
* **Model Version:** gemini-2.5-pro
* **Memory Usage:** 89.5 MB

### Login information

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by 944750720**
## TLDR

Fixes drag-and-drop file functionality for macOS terminal environments. The feature was previously non-functional due to macOS sending focus events followed by individual characters instead of complete file paths, combined with React's asynchronous state updates preventing proper character collection.

## Dive Deeper

This PR addresses a critical issue where drag-and-drop file functionality only worked when the debugger was paused. The root cause was a combination of macOS-specific terminal behavior and React state management challenges:

**Technical Root Cause:**
1. **macOS Drag Event Behavior**: Unlike other platforms, macOS terminal drag operations send a focus event (`\u001b[I`) followed by individual path characters, rather than a complete file path string
2. **React Async State Issue**: `useState` updates are asynchronous, so when the focus event triggered drag mode, subsequent character events arrived before the state update completed, causing character collection to fail
3. **Debugger Timing Effect**: When debugger was paused, the artificial delay allowed async state updates to complete, masking the timing issue

**Solution Implementation:**
- **Dual State Management**: Uses both `useState` (for React rendering) and `useRef` (for immediate access in event handlers) to track drag state
- **Character Collection Mechanism**: Accumulates individual drag characters using functional state updates to avoid closure issues  
- **Smart Timeout Strategy**: 500ms protection timeout for drag mode detection + adaptive completion timeout for character collection
- **Paste Protection**: Ensures normal paste operations (`key.paste = true`) are never interfered with by drag detection logic
- **Focus Event Handling**: Properly processes focus events even when terminal lacks focus to support drag operations

The implementation maintains React best practices while solving the timing-sensitive nature of macOS drag events, ensuring both drag-and-drop and paste functionality work reliably.

## Reviewer Test Plan

**Prerequisites**: macOS system with terminal access

**Primary Functionality Tests**:
1. **Drag-and-Drop Testing**:
   bash
  npm run build && npm start
  
   - Drag file from Finder directly into Gemini CLI terminal
   - Test with various file types: 
.tsx, .js, .py, .md, .json
   - Test with files containing spaces: test file.txt

   - Test with long paths: /Users/username/very/long/path/to/file.tsx

   - Verify paths appear correctly in input prompt

2. **Paste Functionality Verification**:
   - Copy text/code and paste with Cmd+V
   - Verify paste works on first attempt (no double-paste required)
   - Test paste while drag detection might be active

4. **Edge Case Testing**:
   - Press Escape during drag operation (should cancel cleanly)
   - Drag file while typing (should not interfere with existing input)

**Expected Behavior**:
- ✅ Paste operations: Work on first attempt without interference
- ✅ No console errors or state corruption
- ✅ Clean cancellation when interrupted by Escape key
## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ✅  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

**Note**: This fix specifically targets macOS drag behavior. Windows and Linux drag functionality should remain unchanged, but testing is recommended to ensure no regressions.

## Linked issues / bugs

- Fixes #6386 
- Fixes #6449
- Fixes #5729

### CODE REVIEW

File: packages/cli/src/ui/components/InputPrompt.tsx [hunk 1]
```diff
@@ -234,8 +245,95 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
 
   const handleInput = useCallback(
     (key: Key) => {
+      // Terminal drag-and-drop focus event sequences
```
Comment (jacob314) [inline packages/cli/src/ui/components/InputPrompt.tsx]: It is a little bit of a judgment call but I think this should be integrated into KeypressContext.tsx rather than InputPrompt.tsx
Generally that is the right case for any logic dealing with terminal escape and custom escape codes to go. That also has the side benefit of supporting dragging and drop based paste events for everyone listening on input streams rather than just InputPrompt. For example, dragging and dropping a theme.json file would be nice in the theme dialog as a small future feature.

File: packages/cli/src/ui/components/InputPrompt.tsx [hunk 2]
```diff
@@ -234,8 +245,95 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
...
+      const FOCUS_EVENT_IN_EDITOR = '\u001b[I'; // When editor has focus
+      const FOCUS_EVENT_OUT_OF_EDITOR = "'"; // When editor does NOT have focus (macOS Finder.app)
```
Comment (jacob314) [inline packages/cli/src/ui/components/InputPrompt.tsx]: is this right? Seems terrifying that this is the only indication that we are at the end of the event. Seems this might false positive for file paths containing `'`, etc.

File: packages/cli/src/ui/components/InputPrompt.tsx [hunk 3]
```diff
@@ -234,8 +245,95 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
...
+      ) {
+        setDragBuffer((prev) => {
+          const newBuffer = prev + key.sequence;
+
+          // Reset collection timeout for each character
+          if (dragTimeoutRef.current) {
+            clearTimeout(dragTimeoutRef.current);
+          }
+
+          // Set completion timeout - longer delay for long paths
```
Comment (jacob314) [inline packages/cli/src/ui/components/InputPrompt.tsx]: is there really a longer delay for longer paths? doesn't seem like it from the code. would generally recommend a single somewhat longer delay that trying to optimize the delay as this code path should not trigger outside of drag and drop

**Review by jacob314 — COMMENTED** [review]
Great work! Thanks for really digging in and investigating this surprising issue. I'll approve as soon as the code is moved to the suggested location and we add some unittests to avoid future regressions of this functionality. I would suggest we insure the tests validate the baseline case as well as validating that they handle the trickier timeout logic ideally by mocking the passage of time so the tests can still complete almost instantly.

**Comment by jacob314** [comment]
Potential Issues to fix:

   1. `useCallback` Dependencies: The useCallback hook for handleInput includes
      isDragCollecting and dragBuffer in its dependency array. This looks is unnecessary and slightly
      inefficient.
       * isDragCollecting is not used inside the callback; isDragCollectingRef.current is
         used instead.
       * setDragBuffer uses a functional update (setDragBuffer(prev => ...)), which means
         it doesn't need dragBuffer from the closure.
       * Removing these will prevent the handleInput function from being recreated on
         every single character of a dragged file path, improving performance n.

**Comment by 944750720** [comment]
> Great work! Thanks for really digging in and investigating this surprising issue. I'll approve as soon as the code is moved to the suggested location and we add some unittests to avoid future regressions of this functionality. I would suggest we insure the tests validate the baseline case as well as validating that they handle the trickier timeout logic ideally by mocking the passage of time so the tests can still complete almost instantly.

@jacob314 Thank you for your review! I'll check all comments when I'm free and push my modifications. I'm a junior SDE and freshman in React so I'm using AI to assist me to understand issues and solve them.

**Comment by jacob314** [comment]
glad to have you help working on this! please ping me when it is ready for review. heads up that it looks like there are some lint issues to fix. if you check on the failing check you should be able to see what the issues are.

File: packages/cli/src/ui/components/InputPrompt.tsx (hunk 1 above)
Comment (944750720) [reply packages/cli/src/ui/components/InputPrompt.tsx]: I tried to [migrate drag-and-drop logic into KeypressContextu.tsx](https://github.com/google-gemini/gemini-cli/pull/6473/commits/0444b4f03f6c993b17ea94367e09ba4266eb7421). I use `useState` to ensure `newBuffer` safe so that it won't be changed by other event. I'm glad to get advices if it can be more elegent.

File: packages/cli/src/ui/components/InputPrompt.tsx (hunk 2 above)
Comment (944750720) [reply packages/cli/src/ui/components/InputPrompt.tsx]: > is this right? Seems terrifying that this is the only indication that we are at the end of the event. Seems this might false positive for file paths containing ', etc.

image
@jacob314  I've debugged it before. It seems 
`key.sequence==="'" `
when user drags file outside editor(I work in Cursor) to terminal of editor in Macos. But this's really terrible.

**Comment by 944750720** [comment]
> Great work! Thanks for really digging in and investigating this surprising issue. I'll approve as soon as the code is moved to the suggested location and we add some unittests to avoid future regressions of this functionality. I would suggest we insure the tests validate the baseline case as well as validating that they handle the trickier timeout logic ideally by mocking the passage of time so the tests can still complete almost instantly.

Adding unittest.

**Comment by jacob314** [comment]
Ping me when there is a unittest and I'll review again and approve.

**Comment by 944750720** [comment]
> Ping me when there is a unittest and I'll review again and approve.

@jacob314  I've finished unittest. Please kindly check if you have time.

**Comment by 944750720** [comment]
@jacob314 Excuse me. I found GCP cost about 11 dollars because I falsely sync forked google-gemini/main in my fork repo and it ran CI workflows. And I set my API key based on the CONTRIBUTION.md. Is it possible to apply for reimbursement or refund?🥲
https://github.com/944750720/gemini-cli/actions/runs/17191620728

**Comment by jacob314** [comment]
> @jacob314 Excuse me. I found GCP cost about 11 dollars because I falsely sync forked google-gemini/main in my fork repo and it ran CI workflows. And I set my API key based on the CONTRIBUTION.md. Is it possible to apply for reimbursement or refund?🥲 https://github.com/944750720/gemini-cli/actions/runs/17191620728

Sorry you experienced this. Please go to https://cloud.google.com/support/docs/get-billing-support to resolve. Please let me know if that doesn't work.

File: packages/cli/src/ui/contexts/KeypressContext.tsx [hunk 4]
```diff
@@ -78,6 +84,10 @@ export function KeypressProvider({
   const { stdin, setRawMode } = useStdin();
   const subscribers = useRef<Set<KeypressHandler>>(new Set()).current;
 
+  const [, setDragBuffer] = useState('');
```
Comment (jacob314) [inline packages/cli/src/ui/contexts/KeypressContext.tsx]: this is a bit of a code smell. We shouldn't be calling useState if we don't care about using the value in a stateful way. I would use useRef instead as that is probably more aligned with what you want. Could be aligned with isDragCollectingRef

**Comment by 944750720** [comment]
@jacob314 I've resolved conflicts. Please kindly check if you have time

File: docs/cli/authentication.md [hunk 5]
```diff
@@ -19,6 +19,7 @@ The Gemini CLI requires you to authenticate with Google's AI services. On initia
       ```bash
       export GOOGLE_CLOUD_PROJECT="YOUR_PROJECT_ID"
       ```
+
```
Comment (jacob314) [inline docs/cli/authentication.md]: nit: revert this unrelated 1 line change.

**Review by jacob314 — COMMENTED** [review]
Approved once these two hopefully easy to fix comment is addressed.
Can't wait to get this to users!

File: packages/cli/src/ui/contexts/KeypressContext.tsx (hunk 4 above)
Comment (944750720) [reply packages/cli/src/ui/contexts/KeypressContext.tsx]: @jacob314 Thanks! But I faced some questions. Recently I found I can't build project anymore, except with same commit as https://github.com/google-gemini/gemini-cli/pull/6473/commits/fbca51448627402eb89c1381613ea7cabab9c272. I've tried `npm run clean` and `npm install` after removing node_modules/ , but it didn't work. It's so weird... It looks like I imported some wrong code in https://github.com/google-gemini/gemini-cli/pull/6473/commits/d1ed88950649803eeb00a47ec0ee860df971d2ba
The error messages are following: 
```
 ✘ CHJ@MacBookAir  ~/Code/gemini-cli   fix/944750720/fix-drag-file-to-cli-in-macos ±   cd /Users/CHJ/Code/gemini-cli ; /u
sr/bin/env GEMINI_SANDBOX=false PATH=/Users/CHJ/.local/share/fnm/node-versions/v22.15.0/installation/bin:/usr/bin:/bin:/usr/s
bin:/sbin 'NODE_OPTIONS= --require /Applications/Cursor.app/Contents/Resources/app/extensions/ms-vscode.js-debug/src/bootload
er.js  --inspect-publish-uid=http' 'VSCODE_INSPECTOR_OPTIONS=:::{"inspectorIpc":"/var/folders/vs/1mjj6dq531j_zh0x4pkr7rjr0000
gn/T/node-cdp.85343-b5471129-51.sock","deferredMode":false,"waitForDebugger":"","onlyEntrypoint":false,"autoAttachMode":"alwa
ys","fileCallback":"/var/folders/vs/1mjj6dq531j_zh0x4pkr7rjr0000gn/T/node-debug-callback-9d02eaba45dcf765"}' /Users/CHJ/.loca
l/share/fnm/node-versions/v22.15.0/installation/bin/npm run build-and-start 
Using Node v20.19.0
Debugger attached.

> @google/gemini-cli@0.1.21 build-and-start
> npm run build && npm run start

Debugger attached.

> @google/gemini-cli@0.1.21 build
> node scripts/build.js

Debugger attached.
Debugger attached.

> @google/gemini-cli@0.1.21 generate
> node scripts/generate-git-commit-info.js

Debugger attached.
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
Debugger attached.

> @google/gemini-cli@0.1.21 build
> node ../../scripts/build_package.js

Debugger attached.
Debugger attached.
../core/src/core/coreToolScheduler.test.ts:1063:50 - error TS2550: Property 'at' does not exist on type 'any[][]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

1063       const calls = onToolCallsUpdate.mock.calls.at(-1)?.[0] as ToolCall[];
                                                      ~~

../core/src/core/coreToolScheduler.test.ts:1079:64 - error TS2550: Property 'at' does not exist on type 'any[][]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

1079       const completedCalls = onAllToolCallsComplete.mock.calls.at(
                                                                    ~~

../core/src/ide/ide-client.ts:411:9 - error TS2322: Type 'Headers' is not assignable to type 'HeadersInit | undefined'.
  Type 'import("/Users/CHJ/Code/gemini-cli/node_modules/undici/types/fetch").Headers' is not assignable to type 'Headers'.
    The types returned by 'entries()' are incompatible between these types.
      Property '[Symbol.dispose]' is missing in type 'SpecIterableIterator<[string, string]>' but required in type 'HeadersIterator<[string, string]>'.

411         headers: response.headers,
            ~~~~~~~

  ../../node_modules/typescript/lib/lib.esnext.disposable.d.ts:36:5
    36     [Symbol.dispose](): void;
           ~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.dispose]' is declared here.

../core/src/services/chatRecordingService.ts:176:34 - error TS2550: Property 'at' does not exist on type 'MessageRecord[]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

176     return conversation.messages.at(-1);
                                     ~~

../core/src/telemetry/clearcut-logger/clearcut-logger.test.ts:396:21 - error TS2550: Property 'at' does not exist on type 'LogEventEntry[][]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

396       expect(events.at(TEST_ONLY.MAX_EVENTS - 1)).toHaveMetadataValue([
                        ~~

../core/src/tools/glob.ts:9:16 - error TS2305: Module '"glob"' has no exported member 'escape'.

9 import { glob, escape } from 'glob';
                 ~~~~~~

../core/src/tools/glob.ts:146:25 - error TS2352: Conversion of type 'IGlob' to type 'GlobPath[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'IGlob' is missing the following properties from type 'GlobPath[]': length, pop, push, concat, and 28 more.

146         const entries = (await glob(pattern, {
                            ~~~~~~~~~~~~~~~~~~~~~~
147           cwd: searchDir,
    ~~~~~~~~~~~~~~~~~~~~~~~~~
... 
155           signal,
    ~~~~~~~~~~~~~~~~~
156         })) as GlobPath[];
    ~~~~~~~~~~~~~~~~~~~~~~~~~

../core/src/tools/glob.ts:147:11 - error TS2353: Object literal may only specify known properties, and 'cwd' does not exist in type '(err: Error | null, matches: string[]) => void'.

147           cwd: searchDir,
              ~~~

../core/src/tools/grep.test.ts:236:22 - error TS2339: Property 'globStream' does not exist on type 'typeof G'.

236       vi.mocked(glob.globStream).mockRejectedValue(new Error('Glob failed'));
                         ~~~~~~~~~~

../core/src/tools/grep.test.ts:241:22 - error TS2339: Property 'globStream' does not exist on type 'typeof G'.

241       vi.mocked(glob.globStream).mockReset();
                         ~~~~~~~~~~

../core/src/tools/grep.ts:12:10 - error TS2305: Module '"glob"' has no exported member 'globStream'.

12 import { globStream } from 'glob';
            ~~~~~~~~~~

../core/src/tools/read-many-files.ts:12:16 - error TS2305: Module '"glob"' has no exported member 'escape'.

12 import { glob, escape } from 'glob';
                  ~~~~~~

../core/src/tools/read-many-files.ts:217:41 - error TS2345: Argument of type 'any[]' is not assignable to parameter of type 'string'.

217         const entriesInDir = await glob(processedPatterns, {
                                            ~~~~~~~~~~~~~~~~~

../core/src/tools/read-many-files.ts:226:29 - error TS2488: Type 'IGlob' must have a '[Symbol.iterator]()' method that returns an iterator.

226         for (const entry of entriesInDir) {
                                ~~~~~~~~~~~~

src/ui/hooks/useGitBranchName.test.ts:15:32 - error TS2307: Cannot find module 'memfs/lib/volume.js' or its corresponding type declarations.

15 import type { FSWatcher } from 'memfs/lib/volume.js';
                                  ~~~~~~~~~~~~~~~~~~~~~


Found 15 errors.

Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
node:internal/errors:983
  const err = new Error(message);
              ^

Error: Command failed: tsc --build
    at genericNodeError (node:internal/errors:983:15)
    at wrappedFn (node:internal/errors:537:14)
    at checkExecSyncError (node:child_process:882:11)
    at execSync (node:child_process:954:15)
    at file:///Users/CHJ/Code/gemini-cli/scripts/build_package.js:30:1
    at ModuleJob.run (node:internal/modules/esm/module_job:274:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:644:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 29942,
  stdout: null,
  stderr: null
}

Node.js v22.15.0
npm error Lifecycle script `build` failed with error:
npm error code 1
npm error path /Users/CHJ/Code/gemini-cli/packages/cli
npm error workspace @google/gemini-cli@0.1.21
npm error location /Users/CHJ/Code/gemini-cli/packages/cli
npm error command failed
npm error command sh -c node ../../scripts/build_package.js


> @google/gemini-cli-core@0.1.21 build
> node ../../scripts/build_package.js

Debugger attached.
Debugger attached.
src/core/coreToolScheduler.test.ts:1063:50 - error TS2550: Property 'at' does not exist on type 'any[][]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

1063       const calls = onToolCallsUpdate.mock.calls.at(-1)?.[0] as ToolCall[];
                                                      ~~

src/core/coreToolScheduler.test.ts:1079:64 - error TS2550: Property 'at' does not exist on type 'any[][]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

1079       const completedCalls = onAllToolCallsComplete.mock.calls.at(
                                                                    ~~

src/ide/ide-client.ts:411:9 - error TS2322: Type 'Headers' is not assignable to type 'HeadersInit | undefined'.
  Type 'import("/Users/CHJ/Code/gemini-cli/node_modules/undici/types/fetch").Headers' is not assignable to type 'Headers'.
    The types returned by 'entries()' are incompatible between these types.
      Property '[Symbol.dispose]' is missing in type 'SpecIterableIterator<[string, string]>' but required in type 'HeadersIterator<[string, string]>'.

411         headers: response.headers,
            ~~~~~~~

  ../../node_modules/typescript/lib/lib.esnext.disposable.d.ts:36:5
    36     [Symbol.dispose](): void;
           ~~~~~~~~~~~~~~~~~~~~~~~~~
    '[Symbol.dispose]' is declared here.

src/services/chatRecordingService.ts:176:34 - error TS2550: Property 'at' does not exist on type 'MessageRecord[]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

176     return conversation.messages.at(-1);
                                     ~~

src/telemetry/clearcut-logger/clearcut-logger.test.ts:396:21 - error TS2550: Property 'at' does not exist on type 'LogEventEntry[][]'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2022' or later.

396       expect(events.at(TEST_ONLY.MAX_EVENTS - 1)).toHaveMetadataValue([
                        ~~

src/tools/glob.ts:9:16 - error TS2305: Module '"glob"' has no exported member 'escape'.

9 import { glob, escape } from 'glob';
                 ~~~~~~

src/tools/glob.ts:146:25 - error TS2352: Conversion of type 'IGlob' to type 'GlobPath[]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Type 'IGlob' is missing the following properties from type 'GlobPath[]': length, pop, push, concat, and 28 more.

146         const entries = (await glob(pattern, {
                            ~~~~~~~~~~~~~~~~~~~~~~
147           cwd: searchDir,
    ~~~~~~~~~~~~~~~~~~~~~~~~~
... 
155           signal,
    ~~~~~~~~~~~~~~~~~
156         })) as GlobPath[];
    ~~~~~~~~~~~~~~~~~~~~~~~~~

src/tools/glob.ts:147:11 - error TS2353: Object literal may only specify known properties, and 'cwd' does not exist in type '(err: Error | null, matches: string[]) => void'.

147           cwd: searchDir,
              ~~~

src/tools/grep.test.ts:236:22 - error TS2339: Property 'globStream' does not exist on type 'typeof G'.

236       vi.mocked(glob.globStream).mockRejectedValue(new Error('Glob failed'));
                         ~~~~~~~~~~

src/tools/grep.test.ts:241:22 - error TS2339: Property 'globStream' does not exist on type 'typeof G'.

241       vi.mocked(glob.globStream).mockReset();
                         ~~~~~~~~~~

src/tools/grep.ts:12:10 - error TS2305: Module '"glob"' has no exported member 'globStream'.

12 import { globStream } from 'glob';
            ~~~~~~~~~~

src/tools/read-many-files.ts:12:16 - error TS2305: Module '"glob"' has no exported member 'escape'.

12 import { glob, escape } from 'glob';
                  ~~~~~~

src/tools/read-many-files.ts:217:41 - error TS2345: Argument of type 'any[]' is not assignable to parameter of type 'string'.

217         const entriesInDir = await glob(processedPatterns, {
                                            ~~~~~~~~~~~~~~~~~

src/tools/read-many-files.ts:226:29 - error TS2488: Type 'IGlob' must have a '[Symbol.iterator]()' method that returns an iterator.

226         for (const entry of entriesInDir) {
                                ~~~~~~~~~~~~


Found 14 errors.

Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
node:internal/errors:983
  const err = new Error(message);
              ^

Error: Command failed: tsc --build
    at genericNodeError (node:internal/errors:983:15)
    at wrappedFn (node:internal/errors:537:14)
    at checkExecSyncError (node:child_process:882:11)
    at execSync (node:child_process:954:15)
    at file:///Users/CHJ/Code/gemini-cli/scripts/build_package.js:30:1
    at ModuleJob.run (node:internal/modules/esm/module_job:274:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:644:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 30051,
  stdout: null,
  stderr: null
}

Node.js v22.15.0
npm error Lifecycle script `build` failed with error:
npm error code 1
npm error path /Users/CHJ/Code/gemini-cli/packages/core
npm error workspace @google/gemini-cli-core@0.1.21
npm error location /Users/CHJ/Code/gemini-cli/packages/core
npm error command failed
npm error command sh -c node ../../scripts/build_package.js


> @google/gemini-cli-test-utils@0.1.21 build
> node ../../scripts/build_package.js

Debugger attached.
Debugger attached.
Waiting for the debugger to disconnect...
Debugger attached.
Successfully copied files.
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...

> gemini-cli-vscode-ide-companion@0.1.21 build
> npm run compile

Debugger attached.

> gemini-cli-vscode-ide-companion@0.1.21 compile
> npm run check-types && npm run lint && node esbuild.js

Debugger attached.

> gemini-cli-vscode-ide-companion@0.1.21 check-types
> tsc --noEmit

Debugger attached.
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
Debugger attached.

> gemini-cli-vscode-ide-companion@0.1.21 lint
> eslint src

Debugger attached.
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
Debugger attached.
[watch] build started
[watch] build finished
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
node:internal/errors:983
  const err = new Error(message);
              ^

Error: Command failed: npm run build --workspaces
    at genericNodeError (node:internal/errors:983:15)
    at wrappedFn (node:internal/errors:537:14)
    at checkExecSyncError (node:child_process:882:11)
    at execSync (node:child_process:954:15)
    at file:///Users/CHJ/Code/gemini-cli/scripts/build.js:35:1
    at ModuleJob.run (node:internal/modules/esm/module_job:274:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:644:26)
    at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:116:5) {
  status: 1,
  signal: null,
  output: [ null, null, null ],
  pid: 29911,
  stdout: null,
  stderr: null
}

Node.js v22.15.0
Waiting for the debugger to disconnect...
Waiting for the debugger to disconnect...
```

Comment (944750720) [reply packages/cli/src/ui/contexts/KeypressContext.tsx]: > this is a bit of a code smell. We shouldn't be calling useState if we don't care about using the value in a stateful way. I would use useRef instead as that is probably more aligned with what you want. Could be aligned with isDragCollectingRef

@jacob314 I've applied this change before, I remember code didn't work as expection so I want to build and start for test.

**Comment by 944750720** [comment]
@jacob314 I made another PR checkout from commit before importing wrong code: https://github.com/google-gemini/gemini-cli/pull/7174
Please check if [fix(ui): use useRef instead of useState for DragBuffer](https://github.com/google-gemini/gemini-cli/pull/7174/commits/2d2e9d6909d4687b7336f4ce92f0a8f60c39fdaa) is right. This commit made drag-drop doesn't work anymore.

**Comment by SandyTao520** [comment]
Closing this as a newer PR exists.
